### PR TITLE
Documentation for "Configuring Quota Notification Messages"

### LINKED ISSUE #5308: Need to document how to configure Custom Quota Notification messages
Hi,
Need to document how to setup custom quota notification messages:

```
window.OPENSHIFT_CONSTANTS.QUOTA_NOTIFICATION_MESSAGE = {
  // Example quota messages to show in notification drawer
  // "pods": "Upgrade to <a href='http://www.openshift.com'>OpenShift Pro</a> if you need additional resources.",
  // "limits.memory": "Upgrade to <a href='http://www.openshift.com'>OpenShift Online Pro</a> if you need additional resources."
}
```

https://github.com/openshift/openshift-docs/blob/master/install_config/web_console_customization.adoc

Related PR: https://github.com/openshift/origin-web-console/pull/2116

@spadgett

### PR DESCRIPTION
**Pull request by dtaylor113**
Fixes #5308

### CODE REVIEW

File: install_config/web_console_customization.adoc [hunk 1]
```diff
@@ -1014,3 +1014,38 @@ xref:../install_config/install/advanced_install.adoc#advanced-install-configurin
 # See: https://docs.openshift.com/enterprise/latest/install_config/aggregate_logging.html
 #openshift_master_logging_public_url=https://kibana.example.com
 ----
+
+[[configuring-quota-notification-messages]]
+== Configuring Quota Notification Messages
+
+Whenever a user reaches or exceedes a quota a Quota Notification is put into the Notificaiton Drawer.
```
Comment (spadgett) [inline install_config/web_console_customization.adoc]: These aren't really proper names, so I wouldn't capitalize "quota notification" and "notification drawer."

Comment (spadgett) [reply install_config/web_console_customization.adoc]: Would be good to link to the quota and limit ranges topic here.

File: install_config/web_console_customization.adoc [hunk 2]
```diff
@@ -1014,3 +1014,38 @@ xref:../install_config/install/advanced_install.adoc#advanced-install-configurin
...
+A custom quota notification message, per quota type, can be added to the notificaiton.  For example:
+"Your project is over quota.  It is using 200% of 2 cores CPU (Limit). Upgrade to <a href='http://www.openshift.com'>
+OpenShift Pro</a> if you need additional resources.".  The "Upgrade to..." part of the notification is the custom
+message and may contain HTML/links to additional resources.
```
Comment (spadgett) [inline install_config/web_console_customization.adoc]: @ahardin-rh Any suggestions on wording?

File: install_config/web_console_customization.adoc [hunk 3]
```diff
@@ -1014,3 +1014,38 @@ xref:../install_config/install/advanced_install.adoc#advanced-install-configurin
 # See: https://docs.openshift.com/enterprise/latest/install_config/aggregate_logging.html
 #openshift_master_logging_public_url=https://kibana.example.com
 ----
+
+[[configuring-quota-notification-messages]]
+== Configuring Quota Notification Messages
```
Comment (spadgett) [inline install_config/web_console_customization.adoc]: This section should probably be higher up in the doc, maybe under "Configuring Catalog Categories"?

**Review by spadgett — COMMENTED** [review]
Thanks for writing this up @dtaylor113

File: install_config/web_console_customization.adoc [hunk 4]
```diff
@@ -1014,3 +1014,38 @@ xref:../install_config/install/advanced_install.adoc#advanced-install-configurin
...
+
+. Create the following configuration scripts within a file (for example,
+*_quota-messages.js_*):
++
+[source, javascript]
+----
+// Set custom notificaiton messages per quota type/key
+window.OPENSHIFT_CONSTANTS.QUOTA_NOTIFICATION_MESSAGE = {
+    "pods": "Upgrade to <a href='http://www.openshift.com'>OpenShift Pro</a> if you need additional resources.",
+    "limits.memory": "Upgrade to <a href='http://www.openshift.com'>OpenShift Online Pro</a> if you need additional resources."
```
Comment (spadgett) [inline install_config/web_console_customization.adoc]: nit: I'd switch to `'` for the outer strings so the HTML attributes have double quotes, which is more typical.

```
"pods": 'Upgrade to <a href="http://www.openshift.com">OpenShift Pro</a> if you need additional resources.',
"limits.memory": 'Upgrade to <a href="http://www.openshift.com">OpenShift Online Pro</a> if you need additional resources.'
```

**Comment by dtaylor113** [comment]
Thanks @spadgett, I believe I have addressed all of your issues.

File: install_config/web_console_customization.adoc [hunk 5]
```diff
@@ -688,6 +688,51 @@ endif::[]
...
+
+. Create the following configuration scripts within a file (for example,
+*_quota-messages.js_*):
++
+[source, javascript]
+----
+// Set custom notificaiton messages per quota type/key
+window.OPENSHIFT_CONSTANTS.QUOTA_NOTIFICATION_MESSAGE = {
+    "pods": 'Upgrade to <a href="http://www.openshift.com">OpenShift Pro</a> if you need additional resources.',
+    "limits.memory": 'Upgrade to <a href="http://www.openshift.com">OpenShift Online Pro</a> if you need additional resources.'
```
[the remark this answers is not in the document]
Comment (spadgett) [reply install_config/web_console_customization.adoc]: @xiaocwan `memory` refers to memory request and is a different value. It's an alias for `requests.memory`.

See https://docs.openshift.com/container-platform/3.6/admin_guide/quota.html#managed-by-quota

File: install_config/web_console_customization.adoc [hunk 6]
```diff
@@ -688,6 +688,51 @@ endif::[]
 ----
 
 endif::[]
+[[configuring-quota-notification-messages]]
+== Configuring Quota Notification Messages
+
+Whenever a user reaches or exceedes a quota a quota notification is put into the notificaiton drawer.
```
Comment (ahardin-rh) [inline install_config/web_console_customization.adoc]: exceeds a quota, a quota notification

File: install_config/web_console_customization.adoc [hunk 7]
```diff
@@ -688,6 +688,51 @@ endif::[]
...
+A custom quota notification message, per
+xref:../dev_guide/compute_resources.adoc#dev-managed-by-quota[quota resource type], can be added to the notificaiton.  For example:
```
Comment (ahardin-rh) [inline install_config/web_console_customization.adoc]: s/notificaition/notification

File: install_config/web_console_customization.adoc [hunk 8]
```diff
@@ -688,6 +688,51 @@ endif::[]
...
+"Your project is over quota.  It is using 200% of 2 cores CPU (Limit). Upgrade to <a href='http://www.openshift.com'>
+OpenShift Pro</a> if you need additional resources.".  The "Upgrade to..." part of the notification is the custom
+message and may contain HTML/links to additional resources.
```
Comment (ahardin-rh) [inline install_config/web_console_customization.adoc]: HTML or links

**Review by ahardin-rh — COMMENTED** [review]
Just a few minor comments from me. Thanks!

File: install_config/web_console_customization.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (xiaocwan) [reply install_config/web_console_customization.adoc]: @spadgett Sorry but I mean how does the message for "limits.memory" in js show? If user create ResourceQuota with "memory", the message will show when user scale up pod to meet max of ResourceQuota, eg. `oc run myrc --image=aosqe/hello-openshift --limits="cpu=1,memory=1000Mi" --generator=run/v1` . But the "limits.memory" in js does not trigger any message.

[the remark this answers is not in the document]
Comment (xiaocwan) [reply install_config/web_console_customization.adoc]: @spadgett Well, thanks, it works when the quota is defined as
```
apiVersion: v1
kind: ResourceQuota
metadata:
  name: myquota
spec:
  hard:
    cpu: "2"
    limits.memory: 1000Mi
```
Customized message will show up when resource meet max memory:
`# oc run myrc --image=aosqe/hello-openshift --limits="cpu=1,memory=1000Mi" --generator=run/v1`

File: install_config/web_console_customization.adoc (hunk 6 above)
Comment (dtaylor113) [reply install_config/web_console_customization.adoc]: Hi @ahardin-rh, the notification will actually be generated whenever a user reaches a quota or exceeds a quota.  The notification message will be different for each.  I will correct the spelling of "exceeds"  -thanks

File: install_config/web_console_customization.adoc (hunk 7 above)
Comment (dtaylor113) [reply install_config/web_console_customization.adoc]: Hi @ahardin-rh, I'm not sure what you are asking to be addressed here?

File: install_config/web_console_customization.adoc (hunk 8 above)
Comment (dtaylor113) [reply install_config/web_console_customization.adoc]: Hmm...it's not really an 'or' situation.  I'll update to "...may contain HTML such as links to additional resources".

File: install_config/web_console_customization.adoc (hunk 6 above)
Comment (spadgett) [reply install_config/web_console_customization.adoc]: Maybe just "Whenever a user reaches a quota"

File: install_config/web_console_customization.adoc (hunk 7 above)
Comment (spadgett) [reply install_config/web_console_customization.adoc]: @dtaylor113 You have a typo in the spelling of "notification"

**Comment by dtaylor113** [comment]
Hi @ahardin-rh,  I believe I have addressed your review issues  -thanks!

File: install_config/web_console_customization.adoc [hunk 9]
```diff
@@ -688,6 +688,51 @@ endif::[]
 ----
 
 endif::[]
+[[configuring-quota-notification-messages]]
+== Configuring Quota Notification Messages
+
+Whenever a user reaches or exceeds a quota a quota notification is put into the notificaiton drawer.
```
Comment (ahardin-rh) [inline install_config/web_console_customization.adoc]: notification drawer

File: install_config/web_console_customization.adoc [hunk 10]
```diff
@@ -688,6 +688,51 @@ endif::[]
...
+A custom quota notification message, per
+xref:../dev_guide/compute_resources.adoc#dev-managed-by-quota[quota resource type], can be added to the notificaiton.  For example:
```
Comment (ahardin-rh) [inline install_config/web_console_customization.adoc]: the notification

File: install_config/web_console_customization.adoc (hunk 6 above)
Comment (dtaylor113) [reply install_config/web_console_customization.adoc]: Ok, updated.

**Comment by dtaylor113** [comment]
Hi @ahardin-rh,  fixed all spellings of 'notification' ;-)  Thanks!

**Comment by ahardin-rh** [comment]
Thank you!

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../install_config/web_console_customization.adoc#install-config-web-console-customization[Customizing the Web Console]
|Added a xref:../install_config/web_console_customization.adoc#configuring-quota-notification-messages[Configuring Quota Notification Messages] section.
%

**Pull request merged**